torch/distributed/fsdp/symbolic_trace.py

```diff
@@ -0,0 +1,202 @@
+    """
+    tracer: torch.fx.Tracer = torch.fx.Tracer()
+    concrete_args: Optional[Dict[str, Any]] = None
+
+
+class _ExecutionInfo:
+    """
+    Contains the execution order information in the model forward pass.
+
+    Attributes:
+        current_module: record the module that is currently being traced.
+
+        module_forward_order: a list of modules, where the ordering is based on
+            when their forward function is called.
+
+        param_exec_order: a list of parameters ordered based on their execution order.
+
+        _traced_param_set: a set containing all parameters that have been traced.
+
+        module_execution_info_dict: a dict that maps each module to a list of
+            tuple containing a module and a list of named parameters.
+            For a given module, each tuple:
+            1. either contains this module and part of its named_parameters that will be executed together,
+            2. or contains one of its child modules and all of the child module's named_parameters.
+            The list of tuple is ordered based on the parameter execution order.
+    """
+    def __init__(self, root_module: torch.nn.Module) -> None:
+        self.current_module = root_module
+        self.module_forward_order: List[torch.nn.Module] = [root_module]
+
+        named_params_type = List[Tuple[str, torch.nn.Parameter]]
+        self.module_execution_info_dict: Dict[
+            torch.nn.Module,
+            List[Tuple[torch.nn.Module, named_params_type]]
+        ] = dict()
+        self.module_execution_info_dict[root_module] = []
+
+        self.param_exec_order: List[torch.nn.Parameter] = []
+        self._traced_param_set: Set[torch.nn.Parameter] = set()
+
+
+def _patched_create_proxy(
+    create_proxy: Callable,
+    execution_info: _ExecutionInfo,
+    params_dict: Dict[str, torch.nn.Parameter],
+    kind: str,
+    target: torch.fx.node.Target,
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    name: Optional[str] = None,
+    type_expr: Optional[Any] = None
+):
+    """
+    Override of `Tracer.create_proxy` (see
+    https://pytorch.org/docs/stable/fx.html#torch.fx.Tracer.create_proxy).
+    Tracer.create_proxy is called in symbolic tracing for each leaf function/method/module.
+    This override intercepts the recording of each of these operations and
+    update execution_info.module_execution_info_dict.
+
+    Args:
+        create_proxy (Callable):
+            The create_proxy function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to repord the execution information.
+        params_dict (Dict[str, torch.nn.Parameter]):
+            A dict that maps each parameter name to the parameter
+        kind, target, args, kwargs, name, type_expr: inputs to the create_proxy function.
+    """
+    proxy = create_proxy(kind, target, args, kwargs, name, type_expr)
+
+    if kind in ["call_function", "call_method"]:
```

Comment:
Curious, if all you need is getting the execution order, will looping over `GraphModule.graph.nodes` be sufficient?

Reply:
Yes, I believe the flow of "have an input tracer -> trace the forward function and get the execution output -> parse the output" would also work. 

The reason I choose to patch tracer is that we can easily get the execution tree structure (for each module, get the execution order among parameters of this module and its neighboring children module). Below I will use a small example to explain this.

```
           M1
         /     \
        M2     M3
      /    \  /   \
     M4    M5 M6   M7
```

Consider a root module M1 and its inner modules M2-M7. We want to get the execution relation between each module and its children - so get the relation betwen M1, M2, M3, then M2, M4, M5, and M3, M6, M7. This will be used then to schedule parameter rebuild/communication in non-recursive wrapping.
This relation can be easily recorded during the trace since tracing the forward functions will obey this tree structure. 
As to the execution output: this will only record the execution ordering of leaf modules, so M4, M5, M6, M7. We can still use this to infer the execution relation of non-leaf modules, but I think it kind of takes a detour and is less straightforward than just recoding them during the trace.

Reply:
Discussed offline: One caveat of creating a custom tracer class is that since the `input_tracer` can be an instance of any customized class of `nn.Tracer`, we also need to customize other member functions of our tracer so that its `trace()` does the same thing as the `trace()` of the `input_tracer`. 

For now, the design is still patching `create_proxy` and `call_module`, but the patching is only done within a context manager.